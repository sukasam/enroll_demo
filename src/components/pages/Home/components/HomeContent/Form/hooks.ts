import showToast, {
    dismissAllErrorToasts
} from "Components/shared/ShowToaster";
import { getCountryConfig } from "Constants/countryConfig";
import { Alpha2 } from "Constants/countryConfig/enums";
import languages from "Constants/languages";
import { validateSponsor } from "Hydra/validateSponsor";
import { getFirstValidFullName, getFullName } from "Services/utils/fullName";
import isRefValid from "Services/utils/refValidation";
import { useEffect, useState } from "react";

export function useDebounce<T>(value: T, delay: number): T {
    const [debouncedValue, setDebouncedValue] = useState<T>(value);

    useEffect(() => {
        const timer = setTimeout(() => {
            setDebouncedValue(value);
        }, delay);

        return () => {
            clearTimeout(timer);
        };
    }, [value, delay]);

    return debouncedValue;
}

type CountryConfigLanguage = {
    name: string;
    code: string;
};

type AvailableLanguage = {
    label: string;
    value: string;
    defaultCountry: Alpha2;
};

export default function useAvailableLanguages(
    selectedCountry: Alpha2 | null
): AvailableLanguage[] {
    const [availableLanguages, setAvailableLanguages] =
        useState<AvailableLanguage[]>(languages);

    useEffect(() => {
        if (!selectedCountry) {
            setAvailableLanguages(languages);
            return;
        }

        const selectedCountryData = getCountryConfig(selectedCountry, true);
        const countryLanguages = selectedCountryData?.languages?.map(
            (language: CountryConfigLanguage) => ({
                label: language.name,
                value: language.code.toLowerCase(),
                defaultCountry: selectedCountry
            })
        );

        if (!countryLanguages) {
            setAvailableLanguages(languages);
            return;
        }

        setAvailableLanguages(countryLanguages);
    }, [selectedCountry]);

    return availableLanguages;
}

export const checkSponsor = async (
    sponsorId: string,
    enrollerId: string,
    setFormSponsorId: (sponsorId: string) => void,
    setFormSponsorFullName: (sponsorFullName: string) => void,
    setIsLoading: (isLoading: boolean) => void
): Promise<string | boolean> => {
    setIsLoading(true);
    const { isValid, unicityID, refResults } = await isRefValid(sponsorId);
    if (enrollerId === "") {
        return "home_enroller";
    }
    if (!isValid) {
        setFormSponsorFullName("");
        setIsLoading(false);
        return "error_invalid_sponsor";
    }

    const response = await validateSponsor({
        enroller: enrollerId,
        sponsor: unicityID
    });

    if (response.error) {
        setFormSponsorFullName("");
        setIsLoading(false);
        return "error_invalid_sponsor";
    }

    setFormSponsorId(sponsorId);
    setFormSponsorFullName(getFullName(refResults?.humanName));
    setIsLoading(false);
    return true;
};

export const checkRefID = async (
    enrollerId: string,
    setUnicityId: (enrollerId: string) => void,
    setFullName: (enrollerFullName: string) => void,
    setEnrollerAllFullName: (
        enrollerAllFullName: { firstName?: string; lastName?: string } | null
    ) => void,
    setIsLoading: (isLoading: boolean) => void,
    translate: (key: string) => string,
    hiddenEnrollerId?: boolean
): Promise<string | boolean> => {
    setIsLoading(true);
    try {
        const response = await isRefValid(enrollerId);
        const { isValid, refResults, unicityID } = response;

        if (!isValid) {
            setFullName("");
            setIsLoading(false);
            return hiddenEnrollerId
                ? "error_invalid_sponsor"
                : "error_invalid_referrer";
        }

        setUnicityId(unicityID);
        setEnrollerAllFullName(refResults?.humanName || null);
        setFullName(
            getFirstValidFullName([
                refResults?.preferredName,
                refResults?.humanName
            ])
        );
        setIsLoading(false);
        return true;
    } catch (error) {
        setFullName("");
        setIsLoading(false);
        dismissAllErrorToasts();
        showToast(translate("500_error_description"), "error");
        return "";
    }
};
