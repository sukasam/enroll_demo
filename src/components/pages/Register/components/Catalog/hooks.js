import { useCountryConfig } from "Constants/countryConfig";
import { useProducts } from "Contexts/ProductContext";
import { useTranslations } from "Contexts/translation";
import getProducts from "Hydra/getProducts";
import { useEffect } from "react";

export default function useLoadProducts() {
    const { setProducts } = useProducts();
    const { country, language } = useTranslations();
    const countryConfig = useCountryConfig();

    async function load() {
        if (countryConfig?.isNoPurchaseMarket) return;

        const catalog = await getProducts({
            alpha2: country,
            customerType: "Member",
            language
        });
        setProducts(catalog.catalogSlides);
    }

    useEffect(() => {
        load();
    }, [language, country]);
}
