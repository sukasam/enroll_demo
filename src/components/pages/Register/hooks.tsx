import TitleDescription from "Components/shared/TitleDescription";
import {
    ComponentType,
    lazy,
    LazyExoticComponent,
    Suspense,
    useMemo
} from "react";

type Section = {
    id: number;
    title: string;
    type: string;
    props: Record<string, any>;
    sidebarContent: JSX.Element | null;
    completedDescription: JSX.Element | null;
};

const getAccountTitle = (
    isLoggedIn: boolean,
    activeAccordionSection: number | null
): string => {
    if (isLoggedIn && activeAccordionSection !== 1) {
        return "step_account";
    }
    if (isLoggedIn && activeAccordionSection === 1) {
        return "step_update_your_account";
    }
    return "create_account_create_your_account";
};

function LazySidebarContent<T>({
    component: Component,
    sidebarProps
}: {
    component: LazyExoticComponent<ComponentType<T>>;
    sidebarProps?: T;
}): JSX.Element {
    return (
        <Suspense fallback={<div>Loading sidebar...</div>}>
            <Component {...(sidebarProps as any)} />
        </Suspense>
    );
}

export const useSectionsConfig = (
    isLoggedIn: boolean,
    activeAccordionSection: number | null,
    isNoPurchaseMarket = false
): Section[] =>
    useMemo(() => {
        const defaultConfig = [
            {
                id: 1,
                title: getAccountTitle(isLoggedIn, activeAccordionSection),
                type: isLoggedIn ? "UpdateAccount" : "CreateAccount",
                props: {},
                sidebarContent: (
                    <LazySidebarContent
                        component={lazy(
                            () =>
                                import(
                                    "./components/CreateAccount/components/GettingStarted"
                                )
                        )}
                    />
                ),
                completedDescription: (
                    <TitleDescription sectionType="UpdateAccount" />
                )
            },
            {
                id: 2,
                title: "step_starter_kit",
                type: "PackSection",
                props: {},
                sidebarContent: (
                    <LazySidebarContent
                        component={lazy(
                            () =>
                                import(
                                    "./components/Catalog/components/PackBenefits"
                                )
                        )}
                    />
                ),
                completedDescription: (
                    <TitleDescription sectionType="PackSection" />
                )
            },
            {
                id: 3,
                title: "step_shipping",
                type: "ShippingSection",
                props: {},
                sidebarContent: (
                    <LazySidebarContent
                        component={lazy(
                            () => import("Components/shared/ProductSidebar")
                        )}
                    />
                ),
                completedDescription: (
                    <TitleDescription sectionType="ShippingSection" />
                )
            },
            {
                id: 4,
                title: "create_account_agreements",
                type: "DistributorAgreementsSection",
                props: {},
                sidebarContent: (
                    <LazySidebarContent
                        component={lazy(
                            () => import("Components/shared/ProductSidebar")
                        )}
                    />
                ),
                completedDescription: (
                    <TitleDescription sectionType="DistributorAgreementsSection" />
                )
            },
            {
                id: 5,
                title: "step_payment",
                type: "PaymentSection",
                props: {},
                sidebarContent: (
                    <LazySidebarContent
                        sidebarProps={{ showPricingBreakdown: true }}
                        component={lazy(
                            () => import("components/shared/ProductSidebar")
                        )}
                    />
                ),
                completedDescription: (
                    <TitleDescription sectionType="PaymentSection" />
                )
            }
        ];

        const noPurchaseConfig = [
            {
                id: 1,
                title: getAccountTitle(isLoggedIn, activeAccordionSection),
                type: isLoggedIn ? "UpdateAccount" : "CreateAccount",
                props: {},
                sidebarContent: (
                    <LazySidebarContent
                        component={lazy(
                            () =>
                                import(
                                    "./components/CreateAccount/components/GettingStarted"
                                )
                        )}
                    />
                ),
                completedDescription: (
                    <TitleDescription sectionType="UpdateAccount" />
                )
            },
            {
                id: 2,
                title: "step_starter_kit",
                type: "DistributorOnly",
                props: {},
                sidebarContent: (
                    <LazySidebarContent
                        component={lazy(
                            () =>
                                import(
                                    "./components/Catalog/components/PackBenefits"
                                )
                        )}
                    />
                ),
                completedDescription: (
                    <TitleDescription sectionType="PackSection" />
                )
            },
            {
                id: 3,
                title: "step_shipping",
                type: "ShippingSection",
                props: {},
                sidebarContent: null,
                completedDescription: (
                    <TitleDescription sectionType="ShippingSection" />
                )
            },
            {
                id: 4,
                title: "create_account_agreements",
                type: "DistributorAgreementsSection",
                props: {},
                sidebarContent: null,
                completedDescription: (
                    <TitleDescription sectionType="DistributorAgreementsSection" />
                )
            }
        ];

        return isNoPurchaseMarket ? noPurchaseConfig : defaultConfig;
    }, [isLoggedIn, activeAccordionSection, isNoPurchaseMarket]);

export function SectionFactory({
    type,
    props
}: {
    type: string;
    props: Record<string, any>;
}): JSX.Element | null {
    let Component = null;

    switch (type) {
        case "CreateAccount":
            Component = lazy(() => import("./components/CreateAccount"));
            break;
        case "UpdateAccount":
            Component = lazy(() => import("./components/UpdateAccount"));
            break;
        case "PackSection":
            Component = lazy(
                () => import("./components/Catalog/components/PackSection")
            );
            break;
        case "DistributorOnly":
            Component = lazy(
                () => import("./components/Catalog/components/DistributorOnly")
            );
            break;
        case "ShippingSection":
            Component = lazy(() => import("./components/Shipping"));
            break;
        case "DistributorAgreementsSection":
            Component = lazy(
                () => import("./components/DistributorAgreements")
            );
            break;
        case "PaymentSection":
            Component = lazy(() => import("./components/Payment"));
            break;
        default:
            Component = lazy(() => import("./components/CreateAccount"));
    }

    return Component ? (
        <Suspense fallback={<div>Loading...</div>}>
            <Component {...props} />
        </Suspense>
    ) : null;
}
