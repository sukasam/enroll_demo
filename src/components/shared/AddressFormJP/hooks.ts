import { Alpha2 } from "Constants/countryConfig/enums";
import getRegions from "Hydra/getRegions";

export default async function getStates(alpha2: Alpha2): Promise<any> {
    const rawRegions = await getRegions(alpha2);

    return [
        { label: "Select", value: "", disabled: true },
        ...(rawRegions && rawRegions.items
            ? rawRegions.items.map(region => ({
                  label: region.name,
                  value: region.value
              }))
            : [])
    ];
}
