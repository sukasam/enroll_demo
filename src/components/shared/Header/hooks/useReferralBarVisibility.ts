import { useUser } from "Contexts/UserContext";
import { useEffect, useState } from "react";

export default function useReferralBarVisibility(
    referralBar: boolean
): boolean {
    const { enrollerId } = useUser();
    const [showReferralBar, setShowReferralBar] = useState(false);

    useEffect(() => {
        setShowReferralBar(Boolean(enrollerId && referralBar));
    }, [enrollerId, referralBar]);

    return showReferralBar;
}
