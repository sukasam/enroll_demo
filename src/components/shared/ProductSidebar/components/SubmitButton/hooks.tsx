import { useCountryConfig } from "Constants/countryConfig";
import { useOrder } from "Contexts/OrderContext";
import { useProducts } from "Contexts/ProductContext";
import { useUser } from "Contexts/UserContext";
import { useTranslations } from "Contexts/translation";
import { useEffect, useMemo, useState } from "react";

export default function useUserDataMapper(): any {
    const {
        userSignature,
        shipToAddress,
        billingAddress,
        selectedShippingMethod,
        selectedPaymentMethod,
        payPal,
        cashPayment,
        bankWire,
        paymentMethod
    } = useOrder();
    const { shoppingCart } = useProducts();
    const { country } = useTranslations();
    const { userData, userLocale } = useUser();
    const countryConfig = useCountryConfig();

    const [orderData, setOrderData] = useState<any>(null);
    const billingName = useMemo(() => {
        if (billingAddress?.name) {
            const [firstName, ...lastNameParts] = billingAddress.name
                .trim()
                .split(" ");
            return { firstName, lastName: lastNameParts.join(" ") };
        }

        return {
            firstName: userData?.firstName || "",
            lastName: userData?.lastName || ""
        };
    }, [billingAddress, userData]);

    useEffect(() => {
        const cartItems = shoppingCart.map(item => ({
            sku: item.sku,
            qty: 1
        }));

        const address1 = shipToAddress?.address1 || "";
        const address2 = shipToAddress?.address2 || "";
        let city = shipToAddress?.city || "";
        let state = shipToAddress?.state || "";
        const zip = shipToAddress?.zip || "";

        if (shipToAddress?.country === "JP") {
            city = "";
            state = "";
        }

        const data = {
            customer: {
                signature: { value: userSignature || "" },
                status: "Active",
                type: "Associate"
            },
            cartItems,
            payment: {
                firstName: billingName.firstName,
                lastName: billingName.lastName,
                email: userData?.email,
                phone: userData?.phoneNumber,
                address1: billingAddress?.address1 || shipToAddress?.address1,
                address2: billingAddress?.address2 || shipToAddress?.address2,
                city: billingAddress?.city || shipToAddress?.city,
                state: billingAddress?.state || shipToAddress?.state,
                zip: billingAddress?.zip || shipToAddress?.zip,
                country:
                    shipToAddress?.country === "PR"
                        ? "US"
                        : shipToAddress?.country || "",
                method: selectedPaymentMethod,
                ...paymentMethod,
                ...(countryConfig?.marketExceptions.directWorldPay && {
                    bankName: "WorldPay"
                })
            },
            shipToAddress: {
                city,
                country:
                    shipToAddress?.country === "PR"
                        ? "US"
                        : shipToAddress?.country || "",
                state,
                address1,
                address2,
                zip
            },
            shipToName: {
                firstName: shipToAddress?.name,
                lastName: ""
            },
            shipToPhone: userData?.phoneNumber,
            shipToEmail: userData?.email,
            shipMethod: selectedShippingMethod,
            signature: userSignature,
            sourceMarket: country,
            locale: userLocale,
            paypal: payPal || null,
            cashPayment: cashPayment || null,
            bankWire: bankWire || null
        };

        setOrderData(data);
    }, [
        userSignature,
        shipToAddress,
        billingAddress,
        selectedShippingMethod,
        selectedPaymentMethod,
        payPal,
        cashPayment,
        bankWire,
        paymentMethod,
        shoppingCart,
        country,
        userData,
        userLocale,
        billingName
    ]);

    return orderData;
}
