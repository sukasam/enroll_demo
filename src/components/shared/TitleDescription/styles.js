import { css } from "@emotion/react";

export default css`
    font-size: 14px;
    padding-left: 24px;
    display: flex;
    justify-content: space-between;
    align-items: center;

    .icon {
        margin-bottom: -3px;
    }

    .description {
        display: flex;
        align-items: center;

        .text {
            margin-left: 12px;
        }
    }
`;
