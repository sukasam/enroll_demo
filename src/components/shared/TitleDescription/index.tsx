/** @jsxImportSource @emotion/react */
// import { useOrder } from "Contexts/OrderContext";
import { useSectionsConfig } from "Components/pages/Register/hooks";
import { useUser } from "Contexts/UserContext";
import { useMemo } from "react";
import Catalog from "./components/Catalog";
import DistributorAgreements from "./components/DistributorAgreements";
import Payment from "./components/Payment";
import ShippingAddress from "./components/ShippingAddress";
import UpdateAccount from "./components/UpdateAccount";
import styles from "./styles";

type TitleDescriptionProps = {
    sectionType: string;
};

function defalutHooks(): JSX.Element {
    return <div>Title Description sectionType not valid</div>;
}

function useHookFactory(sectionType: string): () => JSX.Element | null {
    switch (sectionType) {
        case "UpdateAccount":
            return UpdateAccount;
        case "PackSection":
            return Catalog;
        case "ShippingSection":
            return ShippingAddress;
        case "DistributorAgreementsSection":
            return DistributorAgreements;
        case "PaymentSection":
            return Payment;
        default:
            return defalutHooks;
    }
}

// Account: {UserFirstName} {UserLastName}, {userEmail}
// StarterKit: Unimate flavor: {Variant/Lemmon} | Ballance pack: {Orange}
// Shipping: {123 Elm Street, Orem, Utah, 84058 Unicted States}
// Distributor Agreements: Signed {Tue 19 Dec 2023 7:22 PM} --- signatureDateTimeStamp
// Payment: {Icon} {selectedPaymentMethod}

export default function TitleDescription({
    sectionType
}: TitleDescriptionProps): JSX.Element {
    const { activeAccordionSection } = useUser();
    const sectionsConfig = useSectionsConfig(true, activeAccordionSection);
    const willRender = useHookFactory(sectionType)();

    const isActiveSection = useMemo(
        () =>
            sectionsConfig.some(
                section =>
                    section.id === activeAccordionSection &&
                    section.type === sectionType // only evaluate the section we are looking for
            ),
        [sectionsConfig, activeAccordionSection, sectionType]
    );

    const shouldRender = useMemo(
        () =>
            (sectionType === "PaymentSection" && isActiveSection) ||
            (!isActiveSection && sectionType !== "PaymentSection"),
        [isActiveSection, sectionType]
    );

    return <div css={styles}>{shouldRender && willRender}</div>;
}
