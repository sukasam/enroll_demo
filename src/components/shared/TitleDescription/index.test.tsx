import { render } from "@testing-library/react";
import { useSectionsConfig } from "Components/pages/Register/hooks";
import { useUser } from "Contexts/UserContext";
import TitleDescription from "./index";

// Mock the necessary hooks
jest.mock("Contexts/UserContext", () => ({
    useUser: jest.fn()
}));

jest.mock("Components/pages/Register/hooks", () => ({
    useSectionsConfig: jest.fn()
}));

jest.mock(
    "./components/Catalog",
    () =>
        function Catalog(): JSX.Element {
            return <div>Catalog Component</div>;
        }
);
jest.mock(
    "./components/DistributorAgreements",
    () =>
        function DistributorAgreements(): JSX.Element {
            return <div>Distributor Agreements Component</div>;
        }
);
jest.mock(
    "./components/Payment",
    () =>
        function Payment(): JSX.Element {
            return <div>Payment Component</div>;
        }
);
jest.mock(
    "./components/DistributorAgreements",
    () =>
        function DistributorAgreements(): JSX.Element {
            return <div>Distributor Agreements Component</div>;
        }
);

jest.mock(
    "./components/Payment",
    () =>
        function Payment(): JSX.Element {
            return <div>Payment Component</div>;
        }
);

jest.mock(
    "./components/ShippingAddress",
    () =>
        function ShippingAddress(): JSX.Element {
            return <div>Shipping Address Component</div>;
        }
);

jest.mock(
    "./components/UpdateAccount",
    () =>
        function UpdateAccount(): JSX.Element {
            return <div>Update Account Component</div>;
        }
);

describe("TitleDescription", () => {
    beforeEach(() => {
        (useUser as jest.Mock).mockReturnValue({
            activeAccordionSection: "someSection"
        });

        (useSectionsConfig as jest.Mock).mockReturnValue([
            { id: "someSection", type: "UpdateAccount" },
            { id: "anotherSection", type: "PackSection" }
        ]);
    });

    it("renders correctly with UpdateAccount sectionType", () => {
        const { asFragment } = render(
            <TitleDescription sectionType="UpdateAccount" />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly with PackSection sectionType", () => {
        const { asFragment } = render(
            <TitleDescription sectionType="PackSection" />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly with ShippingSection sectionType", () => {
        const { asFragment } = render(
            <TitleDescription sectionType="ShippingSection" />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly with DistributorAgreementsSection sectionType", () => {
        const { asFragment } = render(
            <TitleDescription sectionType="DistributorAgreementsSection" />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly with PaymentSection sectionType", () => {
        const { asFragment } = render(
            <TitleDescription sectionType="PaymentSection" />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly with an invalid sectionType", () => {
        const { asFragment } = render(
            <TitleDescription sectionType="InvalidSection" />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
