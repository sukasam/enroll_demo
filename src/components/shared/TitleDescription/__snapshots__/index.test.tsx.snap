// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitleDescription renders correctly with DistributorAgreementsSection sectionType 1`] = `
<DocumentFragment>
  <div
    class="css-jmz7qs"
  >
    <div>
      Distributor Agreements Component
    </div>
  </div>
</DocumentFragment>
`;

exports[`TitleDescription renders correctly with PackSection sectionType 1`] = `
<DocumentFragment>
  <div
    class="css-jmz7qs"
  >
    <div>
      Catalog Component
    </div>
  </div>
</DocumentFragment>
`;

exports[`TitleDescription renders correctly with PaymentSection sectionType 1`] = `
<DocumentFragment>
  <div
    class="css-jmz7qs"
  />
</DocumentFragment>
`;

exports[`TitleDescription renders correctly with ShippingSection sectionType 1`] = `
<DocumentFragment>
  <div
    class="css-jmz7qs"
  >
    <div>
      Shipping Address Component
    </div>
  </div>
</DocumentFragment>
`;

exports[`TitleDescription renders correctly with UpdateAccount sectionType 1`] = `
<DocumentFragment>
  <div
    class="css-jmz7qs"
  />
</DocumentFragment>
`;

exports[`TitleDescription renders correctly with an invalid sectionType 1`] = `
<DocumentFragment>
  <div
    class="css-jmz7qs"
  >
    <div>
      Title Description sectionType not valid
    </div>
  </div>
</DocumentFragment>
`;
